### PR TITLE
fix: allow EVALSHA in ClusterPipeline (#2914)

### LINKED ISSUE #2914: Calling pipelined function evalsha is blocked when running redis in cluster mode
**Version**: 5.0.0

**Platform**: Python 3.10 on MacOS 13

**Description**:

The following code does not work:
```
client = redis.RedisCluster()
pipe = client.pipeline()
pipe.evalsha(sha, 0)
pipe.execute()
``` 

```redis.exceptions.RedisClusterException: ERROR: Calling pipelined function evalsha is blocked when running redis in cluster mode...```

Why is this disallowed? AFAICT this should work fine if the scripts are loaded. It's disallowed in `redis/cluster.py` in `PIPELINE_BLOCKED_COMMANDS`, but no reason is documented.

### PR DESCRIPTION
**Pull request by ahmed5145**
### Description of change

Fixes #2914

`EVALSHA` was listed in `PIPELINE_BLOCKED_COMMANDS`, so `ClusterPipeline.evalsha()` always raised even though `RedisCluster.determine_slot()` already routes `EVAL`/`EVALSHA` by keys (including the 0-key random-slot case).

This removes `EVALSHA` from the blocked list so cluster pipelines can queue `EVALSHA` like other keyed commands. Callers must still:
- load the script on the target node(s) first (e.g. cluster `SCRIPT LOAD`, which loads on all primaries)
- keep all script keys on the same hash slot

Also updates `docs/lua_scripting.rst` and adds regression coverage:
- mock tests that `EVALSHA` is no longer blocked and can be queued on `ClusterPipeline`
- a cluster integration test (`test_evalsha_in_pipeline`) for CI with a live cluster

### Pull Request check-list

- [x] Do tests and lints pass with this change?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [x] Is the new or changed code fully tested?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?

---

> [!NOTE]
> **Medium Risk**
> Changes cluster pipeline command routing and transactional slot selection for Lua scripts; incorrect caller retry after NOSCRIPT could duplicate side effects, though behavior is documented and heavily tested.
> 
> **Overview**
> **Enables `EVALSHA` on sync and async `ClusterPipeline`** by removing it from `PIPELINE_BLOCKED_COMMANDS`, so pipelined scripts can be queued and routed like other keyed commands (same-slot keys, or `numkeys=0` to a random primary).
> 
> **Routing fixes:** Unflagged `EVAL`/`EVALSHA` in cluster pipelines no longer fall through to `COMMAND GETKEYS` (which breaks on Redis &lt;7 when `numkeys=0`); they use keyed default routing instead. **Transactional pipelines** get `_resolve_transaction_slot` so zero-key `EVALSHA` reuses an existing transaction slot (or retargets when the first real keyed command arrives), keeping multi-script / mixed pipelines single-slot without hiding true cross-slot errors.
> 
> **`Script` / `AsyncScript`:** When the client is a `ClusterPipeline`, calls queue `evalsha` directly—no automatic `SCRIPT LOAD` / `NOSCRIPT` retry (callers must reload and retry safely, as documented).
> 
> **Docs** (`lua_scripting.rst`) now describe cluster-pipeline `EVALSHA` behavior, per-node script cache, and `NOSCRIPT` recovery caveats; regression tests cover mocking and live cluster cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6a6c167c892a4b28a18503845f17b303acfba3b5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/test_cluster.py [hunk 1]
```diff
@@ -3835,6 +3836,35 @@ def test_disconnect_calls_disconnect_on_existing_connections(self, r):
...
+        sha = "a" * 40
+        returned = pipe.evalsha(sha, 1, "foo", "bar")
+        assert returned is pipe
+
+        queue = pipe._execution_strategy.command_queue
+        assert len(queue) == 1
+        assert queue[0].args == ("EVALSHA", sha, 1, "foo", "bar")
+
+        assert r.determine_slot("EVALSHA", sha, 1, "foo", "bar") == key_slot(b"foo")
+        zero_key_slot = r.determine_slot("EVALSHA", sha, 0)
```
Comment (Sanjays2402) [inline tests/test_cluster.py]: This zero-key assertion never exercises pipeline execution. `_send_cluster_commands()` calls `_get_command_keys()` before `determine_slot()`, so Redis <7 still fails on `COMMAND GETKEYS EVALSHA ... 0`; this needs an actual zero-key pipeline test or an EVALSHA policy special case.

Comment (ahmed5145) [reply tests/test_cluster.py]: I missed that. Thank you. Aligned the sync path with determine_slot and added a zero-key pipeline execute test (mock + live cluster).

**Review by petyaslavova — CHANGES_REQUESTED** [review]
### Request: make the operational contract explicit in the docs

The routing fix is correct, but the current note understates the operational risk. `EVALSHA` in a `ClusterPipeline` has a materially weaker safety story than non-pipelined `EVALSHA`: the single-node self-healing already documented in this file ("In the event of a NOSCRIPT error, it will load the script and retry executing it") is **not** available in a cluster pipeline — the `Script`/`register_script` fallback and `EVAL` are both disabled there — and the Lua script cache is per-node and not replicated, so any failover, rolling upgrade, resharding, or newly added shard can route the command to a node without the script and raise `NOSCRIPT` with no recovery. The referenced ioredis fix (PR #1201) mitigates this with an automatic pre-`SCRIPT LOAD`; this PR does not, so the caller must own the reload.

Please replace the final line of `docs/lua_scripting.rst` (`Using scripting within pipelines in cluster mode is **not supported**.`) with the following, which matches the file's conventions (plain prose + `` `` `` literals, no cross-ref roles, `.. code:: python`, bold for emphasis):

```rst
``EVALSHA`` can be used inside a ``ClusterPipeline``. Keys must map to the
same hash slot (or ``numkeys`` may be ``0``, in which case the command is
routed to a random primary, exactly as in the non-pipelined case above). In a
transactional pipeline (``pipeline(transaction=True)``), zero-key ``EVALSHA``
reuses the transaction's existing slot when one is already chosen, so multiple
zero-key scripts (or a mix with keyed commands) stay single-slot.

``EVAL``, ``load_scripts``, and ``script_load_for_pipeline`` remain
**not supported** on cluster pipelines.

Important caveats when using ``EVALSHA`` in a cluster pipeline:

- The Lua script cache in Redis is **per-node and is not replicated**.
  ``SCRIPT LOAD`` loads the script onto the *current* primaries only, at the
  moment it is called.
- Any topology change -- a failover that promotes a replica, a rolling
  upgrade that replaces nodes, resharding, or adding a new shard -- can route
  an ``EVALSHA`` to a node whose cache does not contain the script, which
  fails with ``NOSCRIPT`` (``redis.exceptions.NoScriptError``).
- Unlike the non-pipelined ``Script`` object, **a cluster pipeline performs no
  automatic reload or retry** on ``NOSCRIPT``. Recovery is the caller's
  responsibility: catch ``NoScriptError``, re-run ``SCRIPT LOAD``, and
  re-execute the pipeline. Because zero-key ``EVALSHA`` is routed to a random
  primary, the script must be present on **all** primaries.
- In a **transactional** pipeline, a ``NOSCRIPT`` from ``EVALSHA`` is raised at
  ``EXEC`` time and does **not** roll back the other commands (this follows
  Redis ``MULTI``/``EXEC`` semantics), so partial application is possible.

.. code:: python

   >>> from redis.exceptions import NoScriptError
   >>> sha = rc.script_load(lua)  # loads on all current primaries
   >>> def run():
   ...     with rc.pipeline() as pipe:
   ...         pipe.evalsha(sha, 1, "{user}:1")
   ...         return pipe.execute()
   >>> try:
   ...     result = run()
   ... except NoScriptError:
   ...     # a node was missing the script (e.g. after failover/upgrade)
   ...     sha = rc.script_load(lua)  # reload on current primaries
   ...     result = run()
```

**Comment by ahmed5145** [comment]
I updated docs/lua_scripting.rst with the explicit operational contract for EVALSHA in ClusterPipeline (per-node cache, no pipeline reload/retry, transactional NOSCRIPT semantics, and a SCRIPT LOAD retry example).

Please lmk if you have any other feedback!

File: redis/cluster.py [hunk 2]
```diff
@@ -4417,13 +4439,43 @@ def __init__(self, pipe: ClusterPipeline):
...
+        slot_number = self._pipe.determine_slot(*args)
+        if (
+            slot_number is not None
+            and self._pipeline_slots
+            and slot_number not in self._pipeline_slots
+            and not self._transaction_has_keyed_slot
+        ):
+            # Prior slots came only from zero-key EVAL/EVALSHA; retarget.
+            self._pipeline_slots.clear()
+        self._transaction_has_keyed_slot = True
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply redis/cluster.py]: Fixed. _transaction_has_keyed_slot is only set when slot_number is not None (sync + async), with a regression test for a slotless command followed by keyed retargeting.

File: redis/cluster.py [hunk 3]
```diff
@@ -4417,13 +4439,44 @@ def __init__(self, pipe: ClusterPipeline):
...
+            slot_number is not None
+            and self._pipeline_slots
+            and slot_number not in self._pipeline_slots
+            and not self._transaction_has_keyed_slot
+        ):
+            # Prior slots came only from zero-key EVAL/EVALSHA; retarget.
+            self._pipeline_slots.clear()
+        if slot_number is not None:
+            self._transaction_has_keyed_slot = True
+        return slot_number
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply redis/cluster.py:4636]: Skipped. after a keyed WATCH, _transaction_has_keyed_slot is already set, so a later zero-key EVALSHA reuses that slot rather than picking a random one. Retargeting after a prior slot-agnostic command is intentional.

File: redis/cluster.py [hunk 4]
```diff
@@ -4417,13 +4439,44 @@ def __init__(self, pipe: ClusterPipeline):
...
+        if (
+            slot_number is not None
+            and self._pipeline_slots
+            and slot_number not in self._pipeline_slots
+            and not self._transaction_has_keyed_slot
+        ):
+            # Prior slots came only from zero-key EVAL/EVALSHA; retarget.
+            self._pipeline_slots.clear()
+        if slot_number is not None:
+            self._transaction_has_keyed_slot = True
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply redis/cluster.py:4635]: Fixed. zero-key FCALL/FCALL_RO now share the slot-agnostic path with EVALSHA.

File: redis/cluster.py [hunk 5]
```diff
@@ -3641,7 +3655,6 @@ def inner(*args, **kwargs):
     "CONFIG",
     "DBSIZE",
     "ECHO",
-    "EVALSHA",
     "FLUSHALL",
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply redis/cluster.py:3799]: Fixed. AsyncScript no longer awaits ClusterPipeline (that reinitialized and dropped the queued EVALSHA).

File: redis/cluster.py [hunk 6]
```diff
@@ -4417,13 +4439,44 @@ def __init__(self, pipe: ClusterPipeline):
...
+    def _resolve_transaction_slot(self, *args) -> Optional[int]:
+        """
+        Pick a slot for a transactional pipeline command.
+
+        Zero-key EVAL/EVALSHA/FCALL/FCALL_RO can run on any primary. Reuse
+        an existing transaction slot when present so multiple zero-key
+        scripts/functions (or a mix with keyed commands) stay single-slot.
+        """
+        if args[0] in ClusterPipeline.NO_SLOTS_COMMANDS:
+            return None
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply redis/cluster.py:4618]: Yes, this is an intentional defensive guard in the helper. Caller already filters NO_SLOTS_COMMANDS; leaving the check keeps _resolve_transaction_slot safe if called elsewhere. Not changing.

File: redis/cluster.py [hunk 7]
```diff
@@ -3625,11 +3625,11 @@ def inner(*args, **kwargs):
 
 def is_zero_key_eval_command(*args) -> bool:
     """
-    True for EVAL/EVALSHA with numkeys=0 (may run on any primary).
+    True for EVAL/EVALSHA/FCALL/FCALL_RO with numkeys=0 (any primary).
     """
     if len(args) < 3:
         return False
-    if str(args[0]).upper() not in ("EVAL", "EVALSHA"):
+    if str(args[0]).upper() not in ("EVAL", "EVALSHA", "FCALL", "FCALL_RO"):
```
Comment (petyaslavova) [inline redis/cluster.py]: Why have you added FCALL and FCALL_RO - with these covered, the name of this function is no longer appropriate... --> is_zero_key_**eval**_command
Also the whole PR has been focused only on EVAL and EVALSHA so far, can you please share the reasoning why you decided to extend the logic?

**Review by petyaslavova — CHANGES_REQUESTED** [review]
Thanks for turning this around so quickly!

Few more things before we can merge:

1. Please add a regression test (sync and async) that a zero-key EVALSHA followed by two keyed commands on *different* slots still raises CrossSlotTransactionError at execute(). The new `_pipeline_slots.clear()` retarget is the only thing standing between that case and a silent wrong-node execution, so the invariant needs to be pinned down.
2. The async side has no live-cluster coverage. Please mirror the three onlycluster tests you added to TestClusterPipeline in tests/test_cluster.py (keyed, zero-key, and zero-key transactional) under tests/test_asyncio/. Sync/async test parity is a repository requirement.
3. Including FCALL/FCALL_RO in is_zero_key_eval_command changes existing behavior: FCALL was never in PIPELINE_BLOCKED_COMMANDS, so a zero-key FCALL combined with a keyed command in a transactional pipeline used to raise CrossSlotTransactionError and now succeeds. That is probably the behavior we want, but it is outside #2914's scope and is currently undocumented. Please either narrow the helper to EVAL/EVALSHA or call the FCALL change out in the PR description and in docs/lua_scripting.rst.


Few non-blocking notes: the sentence stating that EVAL remains unsupported on cluster pipelines holds for the sync client (ClusterPipeline.eval() raises) but not for the async one, which has no eval override; and the new sync mock tests are the only module-level test functions in tests/test_cluster.py and carry no topology marker, so they also run in the standalone suite, whereas the async ones are inside an onlycluster class.

File: redis/cluster.py [hunk 8]
```diff
@@ -4417,13 +4439,44 @@ def __init__(self, pipe: ClusterPipeline):
...
+        if (
+            slot_number is not None
+            and self._pipeline_slots
+            and slot_number not in self._pipeline_slots
+            and not self._transaction_has_keyed_slot
+        ):
+            # Prior slots came only from zero-key scripts; retarget.
+            self._pipeline_slots.clear()
+        if slot_number is not None:
+            self._transaction_has_keyed_slot = True
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply redis/cluster.py:4635]: Out of scope for this PR. we intentionally narrowed `is_zero_key_eval_command` to EVAL/EVALSHA per maintainer review above. Zero-key FCALL retargeting can be a follow-up with docs.

File: docs/lua_scripting.rst [hunk 9]
```diff
@@ -112,4 +112,46 @@ The following commands are not supported:
...
+  ``SCRIPT LOAD`` loads the script onto the *current* primaries only, at the
+  moment it is called.
+- Any topology change -- a failover that promotes a replica, a rolling
+  upgrade that replaces nodes, resharding, or adding a new shard -- can route
+  an ``EVALSHA`` to a node whose cache does not contain the script, which
+  fails with ``NOSCRIPT`` (``redis.exceptions.NoScriptError``).
+- Unlike the non-pipelined ``Script`` object, **a cluster pipeline performs no
+  automatic reload or retry** on ``NOSCRIPT``. Recovery is the caller's
+  responsibility: catch ``NoScriptError``, re-run ``SCRIPT LOAD``, and
+  re-execute the pipeline. Because zero-key ``EVALSHA`` is routed to a random
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply docs/lua_scripting.rst]: Good catch. I updated the recovery docs to warn that replaying a multi-command pipeline after NOSCRIPT is only safe when the batch is idempotent/single-command.

File: redis/cluster.py (hunk 7 above)
Comment (ahmed5145) [reply redis/cluster.py]: Thanks for the careful review!

Addressed all of the blocking items:

1. Added a sync + async regression test that a zero-key EVALSHA followed by two keyed commands on different slots still raises CrossSlotTransactionError at execute(). That pins the `_pipeline_slots.clear()` retarget invariant.

2. Mirrored the three live-cluster `TestClusterPipeline` EVALSHA cases under `tests/test_asyncio/` (keyed, zero-key, and zero-key transactional).

3. Narrowed `is_zero_key_eval_command` back to EVAL/EVALSHA only and dropped the FCALL-specific tests. Agree that FCALL/FCALL_RO is outside #2914's scope — better as a follow-up if we want that behavior intentionally, with docs.

On the inline comment: no strong reason to extend the helper beyond EVAL/EVALSHA for this PR. It was an opportunistic consistency tweak with `determine_slot()`'s zero-key FCALL path, but you're right that it changed transactional pipeline behavior and made the helper name misleading. Reverted.

Non-blocking notes:
- Clarified the docs sentence: sync `ClusterPipeline.eval()` is blocked; the async cluster pipeline has no `eval` override.
- Marked the sync module-level mock tests with `@pytest.mark.onlycluster` so they no longer run in the standalone suite.

Also tightened the NOSCRIPT recovery docs to warn that blindly re-executing a multi-command pipeline can duplicate side effects (Redis still runs the rest of a non-transactional batch). Left zero-key FCALL slot-agnostic behavior out of this PR on purpose.

**Review by petyaslavova — CHANGES_REQUESTED** [review]
Thanks for turning this around — all three blocking items from the last round are covered: the cross-slot regression test on both stacks, the async live-cluster mirrors, and narrowing the helper back to EVAL/EVALSHA.

One blocker left before merge: in docs/lua_scripting.rst the `>>> try:` / `... except NoScriptError:` lines of the recovery example start at column 0, so they fall outside the `.. code:: python` block and Sphinx renders the example as two disconnected chunks. Please indent those six lines by three spaces to match the lines above them. Docs CI does not build with -W, so it will not catch this for us.

Two optional follow-ups, not merge blockers: the async transaction strategy does not release a connection owned by the previous node when the new retarget changes the transaction slot (the sync path does, redis/cluster.py `_get_client_and_connection_for_transaction`), and a sync counterpart to test_async_script_queues_evalsha_on_cluster_pipeline would complete the parity pair. Happy to take either as a separate issue.

**Comment by ahmed5145** [comment]
Thanks! I fixed the indentation. The six >>> try: / ... except NoScriptError: lines now sit at three spaces, matching the rest of the block, so the whole example renders as one code:: python block.

I also fixed the check-spelling CI failure while I was in there. docs/lua_scripting.rst is the only spellcheck-scanned file this PR touches, and resharding and idempotent were the only new prose words aspell didn't already know, so I added them to .github/wordlist.txt rather than rewording the text you'd already reviewed.

On the Validate building and installing the package (tar.gz) failure: I don't think it's related to this PR. The whl job in the same matrix runs the identical suites via install_and_test.sh and passed in 21 minutes, while tar.gz ran 82 minutes before failing; same code, same tests, so it looks like the docker/test environment hung rather than a real regression. Could you re-run it if it doesn't clear on the new commit?

On the two optional follow-ups, the async transaction connection release and the sync counterpart to test_async_script_queues_evalsha_on_cluster_pipeline,  I agree they're worth doing, and I'd be happy to pick them up, but I'd prefer to keep them out of this PR. Could you open them as a separate issue (or two)? Happy to take them from there.

File: redis/asyncio/cluster.py [hunk 10]
```diff
@@ -3074,6 +3075,8 @@ def __init__(self, pipe: ClusterPipeline) -> None:
         self._explicit_transaction = False
         self._watching = False
         self._pipeline_slots: Set[int] = set()
+        # True once a keyed (non-slot-agnostic) command has fixed the slot
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply redis/asyncio/cluster.py:3078]: False positive. The sync and async fallbacks call different helpers on purpose.

Sync ClusterPipeline uses _get_command_keys() (COMMAND GETKEYS), which breaks on Redis <7 when numkeys=0, hence the explicit EVAL/EVALSHA branch this PR added. The async PipelineStrategy._execute goes through _determine_slot(), which already special-cases EVAL/EVALSHA client-side (redis/asyncio/cluster.py:996): a keyed EVALSHA resolves to keyslot(key) + DEFAULT_KEYED, and only numkeys=0 returns a random slot. Duplicating the branch on the async side would be dead code.

The one real point is thatget_default_node() gate could route keyless if it were ever falsy and that is pre-existing, identical at all four fallback sites, and not reachable here (initialize always sets a primary before _execute). @petyaslavova since it's pre-existing and broader than EVALSHA, I'd fold it into the optional follow-ups as a separate issue rather than this PR.

File: redis/commands/core.py [hunk 11]
```diff
@@ -11003,10 +11003,15 @@ async def __call__(
         args = tuple(keys) + tuple(args)
         # make sure the Redis server knows about the script
         from redis.asyncio.client import Pipeline
+        from redis.asyncio.cluster import ClusterPipeline
```
[the remark this answers is not in the document]
Comment (ahmed5145) [reply redis/commands/core.py:11011]: Fixed

**Comment by ahmed5145** [comment]
fixed what cursor bugbot flagged. sync Script.__call__ now short-circuits ClusterPipeline the same way AsyncScript does, so we queue EVALSHA and never call blocked script_load on NoScriptError. 

I also added test_script_queues_evalsha_on_cluster_pipeline for the sync side. 

Also fixed spellcheck (pipelined → wordlist).

**Review by petyaslavova — COMMENTED** [review]
Testing if GH will show the allow CI run button if a new update is added to the PR ...

**Review by petyaslavova — APPROVED** [review]
LGTM.
